core/coreobjects/include/coreobjects/property_object_impl.h

```diff
@@ -102,6 +102,74 @@ class GenericRecursiveConfigLockGuard : public RecursiveConfigLockGuard
     std::lock_guard<TMutex> lock;
 };
 
+class PropertyUpdateStack
+{
+    struct PropertyUpdateStackItem
+    {
+        PropertyUpdateStackItem(const BaseObjectPtr& value)
+            : value(value)
+            , stackLevel(1)
+        {
+        }
+
+        BaseObjectPtr value;
+        size_t stackLevel;
+    };
+
+public:
+    PropertyUpdateStack() = default;
+
+    // if update was not registered, returns 0, otherwise current stack level
+    size_t registerPropertyUpdating(const std::string& name, const BaseObjectPtr& value)
+    {
+        auto [it, inserted] = updatePropertyStack.try_emplace(name, value);
+        auto& propertItem = it->second;
+
+        if (inserted)
+            return propertItem.stackLevel;
+
+        if (value == propertItem.value)
+            return 0;
+        
+        propertItem.value = value;
+        return ++(propertItem.stackLevel);
+    }
+
+    inline bool isTopStackLevel(const std::string& name, size_t stackLevel) const
+    {
+        return updatePropertyStack.at(name).stackLevel == stackLevel;
+    }
+
+    inline bool isBaseStackLevel(size_t stackLevel) const 
+    {
+        return stackLevel == 1;
+    }
+
+    inline bool isPropertyUpdateRegistered(size_t stackLevel) const
+    {
+        return stackLevel;
+    }
+
+    inline void finalizePropertyUpdating(const std::string& name, size_t stackLevel)
+    {
+        if (stackLevel == 1)
+            updatePropertyStack.erase(name);
+    }
```

Comment:
These should be refactored to not require the `stackLevel` as an input parameter. There should be a `pop` or `unregister` method that returns an enum on whether it was popped at the top, or if it was the last one to pop (stacklevel == 1). If it was the latter, finalize should be called automatically.